File: src/levels/L20-cqrs/ts/cqrs.ts
// Escribir y leer por caminos distintos.
class PlaceOrder {
  constructor(private events: EventBus) {}

  // Lado de escritura: valida reglas del negocio, una orden a la vez.
  // region: PlaceOrder#call
  call(command: PlaceOrderCommand) {
    if (!command.isValid()) throw new InvalidOrder() // reglas del negocio, solo en el lado de escritura
    const order = Order.create(command.attributes)
    this.events.publish('order_placed', order)
  }
  // endregion
}

// region: BoardProjection#on_order_placed
class BoardProjection {
  constructor(private boardView: BoardView) {}

  // Cada cambio actualiza una vista ya lista para leer.
  onOrderPlaced(order: Order) { this.boardView.upsert(order.summary) }
}
// endregion

// region: BoardQuery#call
class BoardQuery {
  constructor(private boardView: BoardView) {}

  // Lado de lectura: una tabla plana, sin joins ni bloqueos. Rápida y sin fila.
  call() { return this.boardView.all() }
}
// endregion

File: src/levels/L20-cqrs/cqrs.rb
# Escribir y leer por caminos distintos.
class PlaceOrder
  # Lado de escritura: valida reglas del negocio, una orden a la vez.
  # region: PlaceOrder#call
  def call(command)
    raise InvalidOrder unless command.valid? # reglas del negocio, solo en el lado de escritura
    order = Order.create!(command.attributes)
    @events.publish(:order_placed, order)
  end
  # endregion
end

# region: BoardProjection#on_order_placed
class BoardProjection
  # Cada cambio actualiza una vista ya lista para leer.
  def on_order_placed(order) = @board_view.upsert(order.summary)
end
# endregion

# region: BoardQuery#call
class BoardQuery
  # Lado de lectura: una tabla plana, sin joins ni bloqueos. Rápida y sin fila.
  def call = @board_view.all
end
# endregion
